src/ralph/parsers.py

```diff
@@ -0,0 +1,86 @@
+"""
+Ralph tracking logs parsers.
+"""
+
+import logging
+import os
+
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+
+
+class BaseParser:
+    """Base tracking logs parser."""
+
+    def parse(self, input_file, filters=None, chunksize=1):
+        """Parse GELF formatted logs (one json string event per row).
+
+        Args:
+            input_file (string): Path to the log file to parse.
+            filters (list): A list of functions to apply on parsed results to
+                filter them.
+            chunksize (int): The amount of log records to process at a time.
+
+        Returns:
+            Parser implementation should return a pandas DataFrame object with
+            all parsed and filtered events.
+
+        """
+        raise NotImplementedError
+
+
+class GELFParser(BaseParser):
+    """GELF formatted logs parser.
+
+    Documentation of the GELF log format can be found in the Graylog project's
+    documentation: https://docs.graylog.org/en/latest/pages/gelf.html
+    """
+
+    def parse(self, input_file, filters=None, chunksize=50000):
+        """Parse GELF formatted logs (one json string event per row).
+
+        Args:
+            input_file (string): Path to the log file to parse (could be
+                gunzipped).
+            filters (list): A list of functions to apply on parsed results to
+                filter them.
+            chunksize (int): The amount of log records to process at a time. A
+                value between 10.000 and 100.000 seems to be a reasonnable
+                choice to parse 1.5M records in a few minutes (typically 2 or 3
+                minutes on a modern computer).
+
+        Returns:
+            A pandas DataFrame object with all parsed and filtered events.
+
+        """
+        logger.info("Parsing: %s", input_file)
+
+        if not os.path.exists(input_file):
+            msg = "Input GELF log file '%s' does not exist"
+            logger.error(msg, input_file)
+            raise OSError(msg % (input_file))
+
+        filters = filters or []
+        events = pd.DataFrame()
+
+        chunks = pd.read_json(input_file, lines=True, chunksize=chunksize)
+        for chunk in chunks:
+            records = pd.read_json("\n".join(chunk["short_message"]), lines=True)
+            logger.debug("Records before filtering: %d", len(records))
+
+            for _filter in filters:
+                logger.info("Current filter: %s", _filter.__name__)
+                if records.empty:
+                    logger.warning(
+                        "Current chunk contains no records, will stop filtering."
+                    )
+                    break
+                records = _filter(records)
+                logger.debug("Filter: %s (records: %d)", _filter.__name__, len(records))
+
+            events = events.append(records, ignore_index=True, sort=False)
```

Comment:
Suggestion : Remove the `empty` filter and add a test here :

```suggestion
            if not records.empty:
                events = events.append(records, ignore_index=True, sort=False)
```

(Or maybe i missed a use case for this empty filter !)

Reply:
Appending an empty DataFrame to a DataFrame has no effect so I think it's not worth adding such test.

Reply:
Hello, I have a question, you are reading chunk by chunk and then appending all to the events dataframe - will it not result in loading almost the full file in memory (filtered lines removed)?

Reply:
Good catch! Yes it will.. even if it's highly optimized in pandas, for this first implementation, we will need to: i. inspect memory usage, and ii. split input files to mitigate the (potential) issue as a first naive approach. We discussed internally of what could be the second implementation using stream processing approach and reactive programing.

Reply:
:thinking: In the meantime, returning a generator would be more efficient.